### DISCUSSION TITLE
First impressions

### DISCUSSION THREAD

**Discussion by gwww** [wezterm/wezterm#616] [General]
Thanks @wez! What an ambitious project!

I currently use `iTerm2`, but I'm always looking for something better, faster, ... What attracts me to `wezterm` is the configuration; I prefer editable config. In playing, since I discovered `wezterm` on Hacker News, I wanted to share my observations. I can open bugs, feature requests, or what ever you wish. Or, it stops with this note. I'm still appreciative of your energy in creating this and I appreciate your "side project" status on this work!

I'm running nightly from about two nights ago on MacOS 11.1, x86.

#### Tab bar
```
﻿﻿﻿c.enable_tab_bar = true
c.hide_tab_bar_if_only_one_tab = false
```
Tab bar still shows when one tab is open. Bug? (`c.` is the Lua table where I build up my config.)

#### Font selection
I've use this font for years: https://github.com/powerline/fonts/tree/master/Inconsolata-g. I can't get `wezterm` to recognize it. It does have a weird style called `g` instead of a more typical style such as bold, italic, etc. Bug? I'm using Fira Code as my backup.

#### Initial delay
When I open `wezterm` and type a command, for example `ls -l`, the first time there is a delay between 1-2 seconds. Here's the output from `time` running `ls -l` twice in a row on a fresh launch of `wezterm`:
```
exa -lh  0.01s user 0.00s system 0% cpu 1.719 total
exa -lh  0.01s user 0.00s system 71% cpu 0.015 total
```
This does not happen all the time. I can reproduce it "sometimes", not sure what condition is triggering it.
#### Resources

I wrote a dump little `bash` script to monitor CPU and memory, to compare against `iTerm2`:

```
#!/usr/bin/env bash

echo "WezTerm  iTerm"
while true; do
  wez=`ps acxo 'pid %cpu rss command'|grep wezterm|tr -s ' '|cut -d ' ' -f 2-3`
  iterm=`ps acxo 'pid %cpu rss command'|grep iTerm2|tr -s ' '|cut -d ' ' -f 2-3`
  echo "$wez  $iterm"
  sleep 1
done
```
With both wez and iTerm idle I see this:
```
0.6 129564  2.0 201924
0.4 129564  1.4 201924
0.4 129564  1.5 201924
0.4 129576  0.8 201924
0.4 129580  1.1 201924
0.4 129580  6.1 201924
0.4 129584  3.7 201924
0.4 129584  2.5 201924
```
CPU is always lower. Memory seemed to vary (although not seen is test run data above).

When running `tree` in `wezterm` and `iTerm` idle:
```
89.2 141700  0.3 202756
96.6 151132  0.8 202756
74.0 158580  1.5 202756
101.7 166932  2.1 202756
159.9 403760  2.5 202756
96.0 254852  0.4 202756
101.3 260220  0.4 202756
99.4 262836  0.5 202756
88.9 267988  0.2 202756
68.5 271968  0.2 202756
75.6 276280  0.9 202756
93.6 288556  0.9 202756
95.8 302744  0.1 202756
46.6 311212  0.3 202756
5.7 311280  4.9 202756
```
When running `tree` in `iTerm` and `wezterm` idle:
```
0.4 311880  5.9 202572
0.3 311892  125.8 202968
0.2 311904  148.7 203564
0.2 311916  151.5 203548
0.3 311920  151.5 203368
0.3 311928  149.2 203356
0.3 311928  127.2 203332
0.2 311932  137.7 203840
0.2 311940  144.6 204080
0.3 311944  134.0 203812
0.2 311956  133.8 203740
0.2 311972  136.6 203700
0.2 311976  126.9 203376
0.2 311976  96.3 203320
0.4 311976  14.5 203260
0.4 311980  4.7 203180
```

The interesting part is that `iTerm` finishes quicker.
```
tree on wezterm:  2.46s user 2.98s system 29% cpu 18.729 total
tree on iTerm2:   2.77s user 3.13s system 43% cpu 13.627 total
```
#### Bitrot
I had a case where I did a `Cmd +` and `Cmd -` (to change font sizes). `wezterm` was running full screen. Bits were left at the bottom of the screen that could not be removed. Restart was the only way to fix. I'm still trying to reproduce so I don't have more details.

#### Reload
It would be nice to have some indicator that a reload has happened. It is not clear when I do a save of my config whether a reload has happened or not.

#### Hide mouse cursor
Mouse cursor turns off when mouse idle and typing.

#### Quit prompt
Would be nice to not have to use mouse. Default OK by pressing Enter would be nice.

**Comment by wez** [comment]
Thanks for the feedback!  I'll take a crack at responding to some of these points inline here, but it may make sense to promote one or two of the things you mentioned to separate issues!

### Tab Bar

> `c.hide_tab_bar_if_only_one_tab = false`
> Tab bar still shows when one tab is open

Is that a transcription error? `hide_tab_bar_if_only_one_tab=true` is the config setting to make the tab bar hide when there is only a single tab.

### Initial delay

It's not clear to me which part is delayed; is it the window showing up, or the text you're typing, or the command running and outputting?

Running `WEZTERM_LOG=trace wezterm 2> /tmp/log-file.txt` will capture some diagnostics with timing information that can help to isolate where things are slow.

### Font Selection

Running `WEZTERM_LOG=wezterm_font=trace wezterm` will capture font-specific diagnostics that might help to figure out what's happening there.

### Bit rot

Tracking in https://github.com/wez/wezterm/issues/605

### Reload

Could perhaps show a toast notification when a reload is triggered?  Please file an enhancement issue for this

### Hide Mouse cursor

Please file an enhancement issue for this!

### Quit Prompt

You can use Y and N to ack the prompts via the keyboard!

**Reply by gwww** [reply]
Doh - on the tab bar hiding.

I can't seem to replicate the performance lag on startup. Very intermittent... when I see it again will try and capture log.

Filed bug for font problem. Filed feature requests for hide mouse cursor and reload notification.

Thanks!!!

**Comment by bew** [comment]
About quit prompt, is this #580  ?
